src/vertex-ai-api.ts

```diff
@@ -33,28 +33,102 @@ interface GeminiRequest {
     mimeType: string;
   };
 }
-
+/**
+ * Custom error class for Gemini API call failures.
+ */
+export class GeminiApiCallError extends Error {}
+/**
+ * Custom error class for image generation API call failures.
+ */
+export class ImageGenerationApiCallError extends Error {}
+/**
+ * Custom error class for JSON parsing errors.
+ */
+export class JsonParseError extends Error {}
+/**
+ * Main interface for the VertexAI api
+ */
 export class VertexAiApi {
-  private readonly _apiEndpoint: string;
-  private readonly _projectId: string;
-  private readonly _baseOptions: GoogleAppsScript.URL_Fetch.URLFetchRequestOptions;
-  readonly VISION_API_LIMIT = 4;
-
-  constructor(apiEndpoint: string, projectId: string) {
-    this._apiEndpoint = apiEndpoint;
-    this._projectId = projectId;
-    this._baseOptions = {
+  /**
+   * Default options for UrlFetchApp.fetch(...) with the auth token
+   */
+  private readonly _baseOptions: GoogleAppsScript.URL_Fetch.URLFetchRequestOptions =
+    {
       method: 'post',
       contentType: 'application/json',
       muteHttpExceptions: true,
       headers: {
         Authorization: 'Bearer ' + ScriptApp.getOAuthToken(),
       },
     };
+  /**
+   * Limit on the image generation api side
+   */
+  readonly IMAGE_GENERATION_API_LIMIT = 4;
+  /**
+   * Creates a new client for interacting with Google Cloud's Vertex AI models.
+   *
+   * @constructor
+   *
+   * @param {string} _projectId - Your Google Cloud Project ID. This is required.
+   * @param {string} [_region='us-central1'] - The region where your AI resources are located (e.g., 'us-central1', 'europe-west4'). Defaults to 'us-central1'.
+   * @param {string} [_apiEndpoint='aiplatform.googleapis.com'] - The base API endpoint for Gemini. Usually, you won't need to change this.
+   * @param {string} [_geminiModel='gemini-1.5-flash'] - The specific Gemini model for text generation tasks (e.g., 'gemini-1.5-flash'). Defaults to the latest flash model.
+   * @param {string} [_imageGenerationModel='imagegeneration'] - The model for image generation tasks.
+   */
+  constructor(
+    private _projectId: string,
+    private _region = 'us-central1',
+    private _apiEndpoint = 'aiplatform.googleapis.com',
+    private _geminiModel = 'gemini-1.5-flash',
+    private _imageGenerationModel = 'imagegeneration'
+  ) {}
+  /**
+   * Constructs the API endpoint URL for the specified Gemini model.
+   *
+   * @protected
+   * @param {string} model - The name of the Gen AI model (e.g., 'gemini-1.5-flash', 'imagegeneration').
+   * @param {string} suffix - The model suffix. Can be found in the end point after ':' (e.g., 'generateContent', 'predict').
+   * @returns {string} The complete API endpoint URL for making requests to the model.
+   */
+  protected getEndPoint(model: string, suffix: string) {
+    return (
+      `https://${this._region}-${this._apiEndpoint}/v1/projects/` +
```

Comment:
I see the endpoint format is different from the previous one - having the region appear twice, is this on purpose?

Reply:
The current uri format includes it twice, e.g. `https://us-central1-aiplatform.googleapis.com/v1/projects/${PROJECT_ID}/locations/us-central1/publishers/google/models/${MODEL_ID}:streamGenerateContent`